docs: add self-hosted LangSmith disaster recovery guide (#4063)
langchain-ai/docs: src/langsmith/self-host-disaster-recovery.mdx

---
title: Disaster recovery for self-hosted LangSmith
sidebarTitle: Disaster recovery
---

> This page describes how to plan, configure, and operate disaster recovery (DR) for a self-hosted LangSmith observability and evaluation deployment. It covers what data must be protected, where it lives, how to back it up, and how to recover the platform after a regional or zonal failure.

Reviewer: ```suggestion
This page describes how to plan, configure, and operate disaster recovery (DR) for self-hosted LangSmith Observability and Evaluation. It covers what data must be protected, where it lives, how to back it up, and how to recover the platform after a regional or zonal failure.
```

<Note>
**Shared responsibility.** For self-hosted deployments you are responsible for backups, replication, restore testing, and recovery procedures for every component, including LangSmith pods and all backing data stores. LangChain is responsible only for the LangSmith software itself. For the equivalent SaaS responsibilities, see the [Shared responsibility model](/langsmith/shared-responsibility-model).
</Note>

<Tip>

> If you have not already, read [Scalability and resilience](/langsmith/scalability-and-resilience) for the architectural primitives (stateless services, queue heartbeats, exactly-once semantics) that this page assumes.

Reviewer: ```suggestion
For details on the architectural primitives (stateless services, queue heartbeats, exactly-once semantics) that this page assumes, refer to [Scalability and resilience](/langsmith/scalability-and-resilience).
```

</Tip>

## What you are recovering

> A self-hosted LangSmith deployment is composed of stateless services backed by four state stores. Recovery planning is almost entirely about the state stores. The stateless services can be recreated at any time by reapplying the Helm chart.

Reviewer: ```suggestion
Self-hosted LangSmith is composed of stateless services backed by four state stores. Recovery planning is almost entirely about the state stores. You can recreate the stateless services at any time by reapplying the Helm chart.
```

> | Layer | Components | State? | Recovery action |

Reviewer: ```suggestion
| Layer | Components | State | Recovery action |
```

|-------|------------|--------|-----------------|
| LangSmith services | `langsmith-frontend`, `langsmith-backend`, `langsmith-platform-backend`, `langsmith-queue`, `langsmith-ingest-queue`, `langsmith-playground`, `langsmith-ace-backend` | Stateless | Reinstall the Helm chart |
| PostgreSQL | Operational data: orgs, workspaces, users, API keys, datasets, prompts, projects, deployments metadata | **Durable** | Restore from backup or replica |
| ClickHouse | Traces and feedback (high volume analytical data) | **Durable** | Restore from backup or replica |
| Blob storage (S3/GCS/Azure Blob) | Run inputs, outputs, errors, manifests, extras, events, attachments (when enabled) | **Durable** | Restore from versioned bucket or replica |
| Redis (or Valkey) | Ephemeral queue state, pub/sub, cache, run heartbeats | Ephemeral | Reprovision; no restore required |
| Kubernetes objects | Helm values, `Secret`s, TLS material, IRSA / Workload Identity bindings | Configuration | Re-apply from source control or back up cluster state |

> <Warning>

Reviewer 2: its okay to not backup redis

All four data stores must be protected. Restoring Postgres without ClickHouse and blob storage (or vice versa) produces an inconsistent installation. References from Postgres to runs in ClickHouse and to objects in blob storage break across the divergence point. Always take coordinated backups, or use point-in-time recovery (PITR) targets that are close together across stores.
</Warning>

## Plan your RPO and RTO

Before designing your DR architecture, define two targets:

- **Recovery Point Objective (RPO):** the maximum amount of data loss your organization can tolerate, measured in time. With managed Postgres PITR, RPO is typically less than 5 minutes. With nightly snapshots only, RPO can be up to 24 hours.
- **Recovery Time Objective (RTO):** the maximum time you can take to restore service after a failure. A warm cross-region replica can deliver RTO in minutes; a cold restore from snapshot can take hours, especially for large ClickHouse datasets.

> The deployment patterns below assume one of three target profiles:

Reviewer: ```suggestion
The following deployment patterns assume one of three target profiles:
```

| Profile | Typical RPO | Typical RTO | Approach |
|---------|-------------|-------------|----------|
| Snapshot-only | 6 to 24 hours | Hours | Daily managed backups of each store. Lowest cost, longest restore. |
| Multi-AZ HA | Seconds | Minutes (zone failure) | Synchronous standby in another AZ for Postgres and ClickHouse, Multi-AZ Redis, zone-redundant blob storage. Standard production posture. |

> | Cross-region DR | Minutes | Minutes to an hour | Asynchronous replication of Postgres, ClickHouse, and blob storage to a second region; standby Kubernetes cluster ready to scale up. Highest cost, fastest recovery from a regional outage. |

Reviewer 2: Typically we recommend backup and restore in a separate region. I'm not sure how well having a passive replica in another region will work.

## Postgres

LangSmith uses PostgreSQL as the primary store for operational and transactional data. **All communication with Postgres uses retries for retry-able errors**, so a brief outage during failover usually does not surface as user-visible errors. A prolonged outage will render the LangSmith API unavailable.

### Use a managed service

> We strongly recommend running Postgres on a managed service in production. Managed services provide automated backups, PITR, and HA failover out of the box. See [Connect external Postgres](/langsmith/self-host-external-postgres) for setup.

Reviewer: ```suggestion
We strongly recommend running Postgres on a managed service in production. Managed services provide built-in automated backups, PITR, and HA failover. For setup, refer to [Connect external Postgres](/langsmith/self-host-external-postgres).
```

### Managed ClickHouse

The fastest path to a resilient ClickHouse is a managed option. See [Connect external ClickHouse](/langsmith/self-host-external-clickhouse).

- **[LangSmith Managed ClickHouse](/langsmith/langsmith-managed-clickhouse):** LangChain operates the ClickHouse cluster, including backups and replication. VPC peering connects it to your LangSmith installation.
- **[ClickHouse Cloud](https://clickhouse.cloud/):** Provides built-in backups, replication, and HA. Available on AWS, GCP, and Azure marketplaces.

### Self-managed replicated cluster

If you self-manage ClickHouse for compliance or air-gap reasons, use a replicated cluster. A single-node ClickHouse instance cannot meet a meaningful RPO.

- Configure a multi-node ClickHouse cluster with replication via [Keeper or ZooKeeper](https://clickhouse.com/docs/architecture/replication).
- Set the `cluster` value in the LangSmith chart so migrations create `Replicated` table engines from the start. **Clustered setups must be configured against a fresh schema**, you cannot convert a standalone instance to clustered later.
- Spread replicas across availability zones.

> - Schedule [`BACKUP TABLE` or `BACKUP DATABASE`](https://clickhouse.com/docs/operations/backup) to object storage on a frequency matching your RPO.

Reviewer 2: i might also recommend altinity backup tool

## Reference deployment patterns

### Single region with Multi-AZ HA (recommended baseline)

This is the minimum production posture and protects against zonal failures. It does not protect against a regional outage.

- Kubernetes node pools across at least two availability zones.
- Postgres in Multi-AZ HA mode (RDS Multi-AZ, Cloud SQL HA, or Flexible Server zone-redundant).
- ClickHouse as a managed service, or a 3-node replicated cluster spread across AZs.
- Redis with Multi-AZ failover enabled.
- Blob storage with versioning enabled and a redundancy tier of at least zone-redundant.
- Daily snapshots of every data store retained for at least 7 days.

### Cross-region active/passive DR

This protects against a regional outage. It is significantly more expensive but is the right pattern for tier-1 deployments.

- A second Kubernetes cluster in the DR region with the LangSmith Helm chart installed but scaled to a low replica count (warm) or zero (cold).
- Postgres cross-region replica (RDS or Aurora cross-region replica, Cloud SQL cross-region replica, Azure Flexible Server cross-region replica). Promote on failover.

> - ClickHouse Cloud or LangSmith Managed ClickHouse with a region failover plan, **or** a replicated self-managed ClickHouse cluster in the DR region with backups copied from the primary.

Reviewer 2: i dont know how well clickhouse replication works cross region. their cloud doesnt really offer this solution either
